packages/node-core/src/api.service.ts

```diff
@@ -2,43 +2,81 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import {NetworkMetadataPayload} from './events';
-import {ISubqueryProject} from './indexer';
+import {ApiConnectionError, ApiErrorType, ConnectionPoolService, ISubqueryProject} from './indexer';
 import {getLogger} from './logger';
 
 const logger = getLogger('api');
 
 const MAX_RECONNECT_ATTEMPTS = 5;
 
-type FetchFunction<T> = (batch: number[]) => Promise<T[]>;
-type FetchFunctionProvider<T> = () => FetchFunction<T>;
+type FetchFunction<T, A> = (batch: number[], api: A) => Promise<T[]>;
+type FetchFunctionProvider<T, A> = () => [FetchFunction<T, A>, A];
 
-export abstract class ApiService<P extends ISubqueryProject = ISubqueryProject, A = any, B = any> {
-  constructor(protected project: P) {}
+export interface IApi<A, SA, B> {
+  fetchBlocks(heights: number[], ...args: any): Promise<B[]>;
+  safeApi(height: number): SA;
+  unsafeApi: A;
+  handleError?(error: Error): ApiConnectionError;
+  apiConnect?(): Promise<void>;
+  apiDisconnect?(): Promise<void>;
+}
+
+export abstract class Api<A, SA, B> implements IApi<A, SA, B> {
```

Comment:
I don't think this is used

Reply:
ApiPromiseConnection extends this

Reply:
Would it make more sense to just have ApiPromiseConnection implement IApi?